modules/frontend/searchsharding.go

```diff
@@ -289,18 +342,90 @@ func (s *searchSharder) shardedRequests(ctx context.Context, metas []*backend.Bl
 			subR := parent.Clone(ctx)
 			subR.Header.Set(user.OrgIDHeaderName, tenantID)
 
-			q := subR.URL.Query()
-			q.Add(api.URLParamBlockID, blockID)
-			q.Add(api.URLParamStartPage, strconv.Itoa(startPage))
-			q.Add(api.URLParamTotalPages, strconv.Itoa(pagesPerQuery))
+			subR, err := api.BuildSearchBlockRequest(subR, &tempopb.SearchBlockRequest{
+				BlockID:       blockID,
+				StartPage:     uint32(startPage),
+				TotalPages:    uint32(pagesPerQuery),
+				Encoding:      m.Encoding.String(),
+				IndexPageSize: m.IndexPageSize,
+				TotalRecords:  m.TotalRecords,
+				DataEncoding:  m.DataEncoding,
+				Version:       m.Version,
+			})
+
+			if err != nil {
+				return nil, err
+			}
 
-			// adding to RequestURI only because weaveworks/common uses the RequestURI field to
-			// translate from http.Request to httpgrpc.Request
-			// https://github.com/weaveworks/common/blob/47e357f4e1badb7da17ad74bae63e228bdd76e8f/httpgrpc/server/server.go#L48
-			subR.RequestURI = api.PathPrefixQuerier + parent.URL.Path + queryDelimiter + q.Encode()
+			subR.RequestURI = buildUpstreamRequestURI(parent.URL.Path, subR.URL.Query())
 			reqs = append(reqs, subR)
 		}
 	}
 
 	return reqs, nil
 }
+
+// queryIngesterWithin returns a new start and end time range for the backend as well as an http request
+// that covers the ingesters. If nil is returned for the http.Request then there is no ingesters query.
+// if the returned start == the returned end then no further querying is necessary
+func (s *searchSharder) ingesterRequest(ctx context.Context, tenantID string, parent *http.Request, searchReq *tempopb.SearchRequest) (uint32, uint32, *http.Request, error) {
+	now := time.Now()
+	ingesterMin := uint32(now.Add(-s.cfg.QueryIngestersWithinMin).Unix())
```

Comment:
**suggestion**: The logic here is a bit hard to follow.  The intent as described is:
```
now -> min: ingester search only
min -> max: ingesters and backend
max -> past: backend only
```

In that case there are 2 thoughts:
* It seems `QueryIngestersWithinMin` is not really ingester-related at all, as the ingesters are queried no matter the value.  It controls whether the backend is searched.   I think it might be clearer to have the two params like:
  * `QueryIngestersUntil: 1h`
  * `QueryBackendAfter: 15m`
* Refactor `ingesterRequest` to only involve ingester, and move the backend start/end trimming elsewhere (i.e. new method)